fix(hpa): deprecated api version, bump to v2

sjoukedv opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

autoscaling/v2beta2 HorizontalPodAutoscaler is deprecated in v1.23+, unavailable in v1.26+; use autoscaling/v2 HorizontalPodAutoscaler

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):


-->

fixes #9347

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

n.a.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added Release Notes.

```release-note
Horizontal Pod Autoscaler - deprecated api version, bump to v2

```

longwuyuan commented:
@sjoukedv Thanks for your contribution
- Please check https://github.com/kubernetes/ingress-nginx/pull/8872. I think they are either conflcting on how to solve the problem or they are duplicates.
- We need data here that explaains backwards compatibility in the special way this project handles users. We officially do not support more than current -2 versions of K8s but we also try hard not to break backwards compatibility, to be best possible chances. That means I still did not look at the code change but want to flesh the content here with the design aspects as well as the context of that other PR i linked above

longwuyuan commented:
I just discovered a odd thing that the other PR I linked has 0 files changed https://github.com/kubernetes/ingress-nginx/pull/8872#issuecomment-1330344405 

Maybe we need to wait and find out more. The focus is the conditional use of hpa api version based on K8s version etc.

sjoukedv commented:
> @sjoukedv Thanks for your contribution
> 
> * Please check [Determine the hpa version and select the hpa version #8872](https://github.com/kubernetes/ingress-nginx/pull/8872). I think they are either conflcting on how to solve the problem or they are duplicates.
> * We need data here that explaains backwards compatibility in the special way this project handles users. We officially do not support more than current -2 versions of K8s but we also try hard not to break backwards compatibility, to be best possible chances. That means I still did not look at the code change but want to flesh the content here with the design aspects as well as the context of that other PR i linked above

I can refactor with a helm value that allows you to still use the older (deprecated) version. I'll give it a try a soon.

The other PR has the same purpose, but contains no changed files.

sjoukedv commented:
Okay, I have abstracted the value of the apiVersion, so that people can override it with a different one. I'll leave it to you which one you would like to set as default (now it still ``autoscaling/v2beta2`).

Dry run without the version yields the template as is now:

```
$ helm install test ./charts/ingress-nginx/ --values=./charts/ingress-nginx/values.yaml --set controller.autoscaling.enabled=true --dry-run | grep 'HorizontalPodAutoscaler' -B 1
W1129 11:18:39.432722  123922 warnings.go:70] autoscaling/v2beta2 HorizontalPodAutoscaler is deprecated in v1.23+, unavailable in v1.26+; use autoscaling/v2 HorizontalPodAutoscaler
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
```

Specifying the apiVersion:

```
$ helm install test ./charts/ingress-nginx/ --values=./charts/ingress-nginx/values.yaml --set controller.autoscaling.enabled=true --set controller.autoscaling.apiVersion=autoscaling/v2 --dry-run | grep 'HorizontalPodAutoscaler' -B 1
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
```

longwuyuan commented:
cc @rikatz @tao12345666333

rikatz commented:
Sgtm, when was v2beta2 deprecated and when v2 got stable? I would default to the stable one if it is available for at least 3 releases

sjoukedv commented:
> Sgtm, when was v2beta2 deprecated and when v2 got stable? I would default to the stable one if it is available for at least 3 releases

https://kubernetes.io/docs/reference/using-api/deprecation-guide/#horizontalpodautoscaler-v125

> the autoscaling/v2 API version, available since v1.23

So that makes two releases. Important to note for the upcoming version 1.26

> The autoscaling/v2beta2 API version of HorizontalPodAutoscaler will no longer be served in v1.26.

rikatz commented:
1.23, 1.24, 1.25 = 3 releases :)
let’s move forward. Can you just check the ci errors?
/ok-to-test

sjoukedv commented:
> 1.23, 1.24, 1.25 = 3 releases :) let’s move forward. Can you just check the ci errors? /ok-to-test

I have changed the default apiVersion to `autoscaling/v2` and added documentation about the parameter in the README (this failed the CI before)

sjoukedv commented:
Can someone check the pipeline? I didn't make any changes to the source files yet that is where it fails.

longwuyuan commented:
Did you edit the README.md undr /charts manually to add that one line for autoscaling ?

sjoukedv commented:
> nually to add that one line for autoscaling ?

Yes, I did and afterwards I ran the `helm-docs` go module on it. Perhaps a re-trigger would do the trick already.

longwuyuan commented:
@sjoukedv thanks
/lgtm

wasker commented:
When the chart with this change is going to be published? Long-awaited fix...
